### ISSUE TITLE
5.2.4 symfony/phpunit-bridge changed some direct-dependency deprecation to self one.

### ISSUE THREAD

**Issue by VincentLanglet** [symfony/symfony#40400]
**Symfony version(s) affected**: 5.2.4

**Description**

**How to reproduce**
With 5.2.3 version, Sonata build were green: https://github.com/sonata-project/SonataPageBundle/runs/1952183310
With 5.2.4 version, it doesn't: https://github.com/sonata-project/SonataPageBundle/pull/1284/checks?check_run_id=2048123569

For instance:
```
1x: The "Sonata\PageBundle\Command\BaseCommand" class extends "Symfony\Bundle\FrameworkBundle\Command\ContainerAwareCommand" that is deprecated since Symfony 4.2, use {@see Command} instead.
    1x in BaseCommandTest::setUp from Sonata\PageBundle\Tests\Command
```
Moved from `Remaining direct deprecation notices` to `Remaining self deprecation notices`.

I think it's a bug:
- If I introduce a deprecation in the Sonata code, I expect this to be reported as a self deprecation. And I expect to update all the code/tests. Since I introduce the deprecation, I can provide a BC way to update the code.
- If someone introduce a deprecation in a vendor, this should be a direct-dependency deprecation, this should not be reported as a self one because I extends this final/deprecated class. Because build shouldn't start failing for a version bump, and because public library are not always able to fix the deprecation in a BC way.

Is it related to https://github.com/symfony/symfony/pull/40067 @nicolas-grekas ?

**Comment by nicolas-grekas** [comment]
Duplicate of #40390?

**Comment by VincentLanglet** [comment]
> Duplicate of #40390?

The root cause is maybe (maybe not ?) the same, but I won't call this a duplicate.
#40390 talk about a fatal error. 

I don't have fatal error, just deprecations notices in the wrong (according to me) category.

**Comment by stof** [comment]
Well, extending a deprecated class in your own code is indeed a deprecated usage happening in your code, not in vendor code. This is something that gets fixed by changing your own code. So I would say that the new behavior is the right one here.

> * Because build shouldn't start failing for a version bump, and because public library are not always able to fix the deprecation in a BC way.
It will be a vendor deprecation for projects using SonataPageBundle as a vendor (so that they can accept such deprecations in their build by failing only on self deprecations, as they have to wait until SonataPageBundle fixes the deprecated usage).

if your build fails when you use a deprecated API, it is totally possible that it starts failing when updating dependencies inside a semver range. That's the whole point of deprecations: they are not introduced only in major releases.

**Comment by VincentLanglet** [comment]
> Well, extending a deprecated class in your own code is indeed a deprecated usage happening in your code, not in vendor code. This is something that gets fixed by changing your own code. So I would say that the new behavior is the right one here.
> 
> > * Because build shouldn't start failing for a version bump, and because public library are not always able to fix the deprecation in a BC way.
> >   It will be a vendor deprecation for projects using SonataPageBundle as a vendor (so that they can accept such deprecations in their build by failing only on self deprecations, as they have to wait until SonataPageBundle fixes the deprecated usage).
> 
> if your build fails when you use a deprecated API, it is totally possible that it starts failing when updating dependencies inside a semver range. That's the whole point of deprecations: they are not introduced only in major releases.

If a vendor deprecate the method `foo()` when I use `foo()` in my code, it's reported as a direct dependency deprecation ; not a self one.
If a vendor deprecate the class `Foo` when I extends `Foo` in my code, it was reported as a direct dependency deprecation ; now it's reported as a self one.

1) This change is not consistent.
2) This change is not logic to me, the meaning of `self/direct/...` seems to be the origin of the deprecation, the place where `trigger_deprecation`/`@deprecated` is used. In this case, this is deprecated in the vendor, not in my code.
3) Since a vendor deprecation cannot always be solved without BC-break, it means that with this change, it's now not possible to use "max[self]=0"  for a public repository.

**Comment by VincentLanglet** [comment]
WDYT @nicolas-grekas ?

**Comment by nicolas-grekas** [comment]
Would you be able to have a look? I can't until next week personally.

**Comment by VincentLanglet** [comment]
> Would you be able to have a look? I can't until next week personally.

I took a look but I don't understand a lot of changes.
I would say it's because of this: https://github.com/symfony/symfony/pull/40067/files#diff-12ca211e6c8ad6675e30cda96f80935f0968b05a58813d570d55b81a9dcdd201R85-R115
But you certainly understand these changes better than I am, because you made them ^^'

**Comment by OskarStark** [comment]
@greg0ire I know you are into this topic, can you maybe help us out? 🙏

**Comment by greg0ire** [comment]
I had a look, and the triggering file changes between the 2 versions. In the case that works, it is `/home/greg/dev/SonataPageBundle/vendor/symfony/error-handler/DebugClassLoader.php`. In the case that does not, it is `/home/greg/dev/SonataPageBundle/src/Command/BaseCommand.php`. I will perform a bisection.

UPD: the bisection indicates that the first bad commit is d0eb698b1167bfdf2bca8f1e49993ebff8d9687b

"[PhpUnitBridge] fix reporting deprecations when they come from DebugClassLoader"

That would make sense since it involves `DebugClassLoader`

It does not say what was broken or why the changes fix that though.

In the monorepo, the hash is https://github.com/symfony/symfony/commit/6a4312deaca79de166a70f93280088360a435b2b , and you were right @VincentLanglet , it does come from https://github.com/symfony/symfony/pull/40067

This part is interesting: https://github.com/symfony/symfony/pull/40067/commits/6a4312deaca79de166a70f93280088360a435b2b#diff-12ca211e6c8ad6675e30cda96f80935f0968b05a58813d570d55b81a9dcdd201R74-R78 : it does a special treatment in the case of `DebugClassLoader`

I think that change was done so that the deprecations comes from `BaseCommand`, but that's not really correct is it? In your case, the triggeringFile should be the one holding the deprecation, so the file holding the definition for `Symfony\Bundle\FrameworkBundle\Command\ContainerAwareCommand`. I'm not sure there will be an easy way to get that piece of information though, since it's unlikely to appear in the backtrace :grimacing:

**Comment by nicolas-grekas** [comment]
Fixed in #40508
